File: backend/src/services/optimization.js
function clamp(n, a, b) {
  return Math.max(a, Math.min(b, n));
}

export function localOptimize(input) {
  const demand = input.predictedDemandKg ?? 480;
  const inventory = input.currentInventoryKg ?? 180;
  const safety = input.safetyStockKg ?? 80;
  const priceBuy = input.purchasePriceInr ?? 24;
  const priceSell = input.expectedSellingPriceInr ?? 32.4;
  const spoilage = input.spoilageProb ?? 0.08;
  const capacity = input.storageCapacityKg ?? 4000;
  const budget = input.budgetInr ?? 50000;
  const transport = input.transportCostPerKg ?? 1.2;
  const gap = Math.max(0, demand + safety - inventory);
  const affordable = Math.floor(budget / (priceBuy + transport));
  const qty = Math.min(gap, affordable, Math.max(0, capacity - inventory));
  const expectedProfit = Math.round(
    qty * (priceSell - priceBuy - transport) * (1 - spoilage) - Math.max(0, inventory - demand) * priceBuy * 0.08
  );
  const conf = clamp(0.86 - Math.abs(input.demandUncertainty ?? 0.12) * 0.4, 0.55, 0.93);
  let risk = 'LOW';
  if (spoilage > 0.25 || (input.oversupplyProb ?? 0) > 0.55) risk = 'HIGH';
  else if (spoilage > 0.12 || (input.oversupplyProb ?? 0) > 0.3) risk = 'MODERATE';
  return {
    action: qty > 40 ? 'BUY' : qty === 0 && inventory > demand * 1.15 ? 'HOLD_SELL' : 'HOLD',
    quantityKg: qty,
    expectedProfitInr: expectedProfit,
    confidence: Number(conf.toFixed(2)),
    risk,
    fallback: true,
  };
}

export function buildFactors({ demandDelta, inventoryDelta, priceDelta, supplyDelta, weatherDelta, spoilageDelta }) {
  return [
    { label: 'Demand', impactPct: demandDelta, direction: demandDelta >= 0 ? 'up' : 'down' },
    { label: 'Inventory', impactPct: inventoryDelta, direction: inventoryDelta >= 0 ? 'up' : 'down' },
    { label: 'Expected price', impactPct: priceDelta, direction: priceDelta >= 0 ? 'up' : 'down' },
    { label: 'Supply', impactPct: supplyDelta, direction: supplyDelta >= 0 ? 'up' : 'down' },
    { label: 'Weather impact', impactPct: weatherDelta, direction: weatherDelta >= 0 ? 'up' : 'down' },
    { label: 'Spoilage risk', impactPct: spoilageDelta, direction: spoilageDelta >= 0 ? 'up' : 'down' },
  ];
}

File: ml-service/app/optimization.py
from __future__ import annotations

import numpy as np
from scipy.optimize import minimize_scalar


def expected_profit(q, demand_mu, demand_std, buy, sell, spoilage, transport, inventory):
    q = max(0, q)
    available = inventory + q
    # truncated demand expectation approximation
    sold = min(available, demand_mu)
    unsold = max(0, available - demand_mu)
    lost = max(0, demand_mu - available)
    margin = sell - buy - transport
    profit = sold * margin - unsold * buy * (0.12 + spoilage) - lost * margin * 0.25 - q * transport * 0.15
    # uncertainty penalty
    profit -= demand_std * 0.08 * (q / max(demand_mu, 1))
    return profit


def optimize_procurement(payload: dict) -> dict:
    demand = float(payload.get("predictedDemandKg", 480))
    lo = float(payload.get("demandLower", demand * 0.85))
    up = float(payload.get("demandUpper", demand * 1.15))
    std = max(8, (up - lo) / 2)
    inventory = float(payload.get("currentInventoryKg", 180))
    buy = float(payload.get("purchasePriceInr", 24))
    sell = float(payload.get("expectedSellingPriceInr", payload.get("predictedPriceInr", 32.4)))
    spoilage = float(payload.get("spoilageProb", 0.08))
    transport = float(payload.get("transportCostPerKg", 1.2))
    capacity = float(payload.get("storageCapacityKg", 4000))
    budget = float(payload.get("budgetInr", 50000))
    max_q = min(max(0, capacity - inventory), budget / max(buy + transport, 1))

    def objective(q):
        return -expected_profit(q, demand, std, buy, sell, spoilage, transport, inventory)

    if max_q <= 1:
        q_star = 0.0
    else:
        res = minimize_scalar(objective, bounds=(0, max_q), method="bounded")
        q_star = float(res.x)

    profit = expected_profit(q_star, demand, std, buy, sell, spoilage, transport, inventory)
    over = float(payload.get("oversupplyProb", 0.2))
    conf = float(np.clip(0.9 - std / demand * 0.5 - over * 0.15, 0.55, 0.93))
    if spoilage > 0.28 or over > 0.6:
        risk = "CRITICAL"
    elif spoilage > 0.18 or over > 0.4:
        risk = "HIGH"
    elif spoilage > 0.1 or over > 0.25:
        risk = "MODERATE"
    else:
        risk = "LOW"
    action = "BUY" if q_star > 40 else "HOLD"
    if inventory > demand * 1.2 and q_star < 20:
        action = "HOLD_SELL"
    return {
        "action": action,
        "quantityKg": int(round(q_star)),
        "expectedProfitInr": int(round(profit)),
        "confidence": round(conf, 2),
        "risk": risk,
        "objective": "max expected profit subject to spoilage, capacity, budget, uncertainty",
        "constraints": {"maxQty": int(max_q), "budgetInr": budget, "capacityKg": capacity},
    }
